### DISCUSSION TITLE
Tree-shaking and Typescript

### DISCUSSION THREAD

**Discussion by majelbstoat** [vercel/next.js#11773] [Help]
I'm using Next 9.3.6 and I've arranged my components like:
```ts
  components/
    Header/
      Header.tsx
    Sidebar/
      Sidebar.tsx
    index.ts
```

where index.ts is:
```ts
  export * from './Header/Header.tsx'
  export * from './Sidebar/Sidebar.tsx'
  // ...
```

so that in a page, I can do:
```ts
import { Header, Sidebar } from "../components"
```

There's about 60 components defined, and the overwhelming majority of them aren't used on every page. But when I analyze the bundle I have a very large chunk, shared by all pages, and it appears to contain all my components:

<img width="1680" alt="Screenshot 2020-04-09 07 23 39" src="https://user-images.githubusercontent.com/425787/78890489-c11c4e80-7a33-11ea-8893-1705f2c2f695.png">

my tsconfig.json is:
```json
  "compilerOptions": {
    "allowJs": true,
    "baseUrl": ".",
    "forceConsistentCasingInFileNames": true,
    "esModuleInterop": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "lib": ["dom", "dom.iterable", "esnext"],
    "module": "esnext",
    "moduleResolution": "node",
    "noEmit": true,
    "noFallthroughCasesInSwitch": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "preserveConstEnums": true,
    "removeComments": false,
    "resolveJsonModule": true,
    "skipLibCheck": true,
    "strict": true,
    "target": "es5"
  }
```

and I'm using the `tsconfig-paths-webpack-plugin` to support the `baseUrl` field. I haven't changed the `module` or the `target`.

I have about 10 different pages so far, and I know the bundler probably has a somewhat complex chunking algorithm, but I wanted to check that this is the correct way to import modules so as not to break tree-shaking. My assumption would be I'd have a moderately sized common chunk, and then some smaller page-specific ones, but that doesn't seem to be happening. Am I doing something wrong?

**Comment by steric85** [comment]
It's being discussed here: https://github.com/vercel/next.js/issues/12557
